gensim/models/bm25model.py

```diff
@@ -0,0 +1,391 @@
+        else:
+            pass
+
+    @abstractmethod
+    def precompute_idfs(self, dfs, num_docs):
+        """Precompute inverse term document frequencies, which will be used to weight term frequencies
+        for the documents.
+
+        Parameters
+        ----------
+        dfs : dict of (int, int)
+            A mapping from term ids to term document frequencies.
+        num_docs : int
+            The total number of documents in the training corpus.
+
+        Returns
+        -------
+        idfs : dict of (int, float)
+            A mapping from term ids to inverse term document frequencies.
+
+        """
+        pass
+
+    @abstractmethod
+    def get_term_weights(self, num_tokens, term_frequencies, idfs):
+        """Compute vector space weights for a set of terms in a document.
+
+        Parameters
+        ----------
+        num_tokens : int
+            The number of tokens in the document.
+        term_frequencies : ndarray
+            1D array of term frequencies.
+        idfs : ndarray
+            1D array of inverse term document frequencies.
+
+        Returns
+        -------
+        term_weights : ndarray
+            1D array of vector space weights.
+
+        """
+        pass
+
+    def __getitem__(self, bow):
+        is_corpus, bow = utils.is_corpus(bow)
+        if is_corpus:
+            return self._apply(bow)
+
+        num_tokens = sum(freq for term_id, freq in bow)
+
+        term_ids, term_frequencies, idfs = [], [], []
+        for term_id, term_frequency in bow:
+            term_ids.append(term_id)
+            term_frequencies.append(term_frequency)
+            idfs.append(self.idfs.get(term_id) or 0.0)
+        term_frequencies, idfs = np.array(term_frequencies), np.array(idfs)
+
+        term_weights = self.get_term_weights(num_tokens, term_frequencies, idfs)
+
+        vector = [
+            (term_id, float(weight))
+            for term_id, weight
+            in zip(term_ids, term_weights)
+        ]
+        return vector
+
+
+class OkapiBM25Model(BM25ABC):
+    """The original Okapi BM25 scoring function of Robertson et al. [2]_.
+
+    Examples
+    --------
+    .. sourcecode:: pycon
+
+        >>> from gensim.corpora import Dictionary
+        >>> from gensim.models import OkapiBM25Model
+        >>> from gensim.test.utils import common_texts
+        >>>
+        >>> dictionary = Dictionary(common_texts)  # fit dictionary
+        >>> model = OkapiBM25Model(dictionary=dictionary)  # fit model
+        >>>
+        >>> corpus = [dictionary.doc2bow(line) for line in common_texts]  # convert corpus to BoW format
+        >>> vector = model[corpus[0]]  # apply model to the first corpus document
+
+    References
+    ----------
+    .. [2] Robertson S. E., Walker S., Jones S., Hancock-Beaulieu M. M., Gatford M. (1995).
+       `Okapi at TREC-3 <http://research.microsoft.com/pubs/67649/okapi_trec3.pdf>`_.
+       *NIST SPECIAL PUBLICATION SP*.
```

Comment:
What does this line mean? It's all in caps and in emphasis, so I assume it's important. It it a part of the reference string, or some warning? What does `SP` stand for?

Reply:
That's the name of the collection in which the article was published.

Reply:
OK I googled it up; it's simply this:

```suggestion
       *National Institute of Standard and Technology Special Publication (NIST SP)*.
```